File: android/app/src/main/java/dev/nytweetdeck/android/ui/PostTextSegments.kt
package dev.nytweetdeck.android.ui

import java.net.URI

internal sealed interface PostTextSegment {
    val text: String

    data class Plain(override val text: String) : PostTextSegment

    data class Hashtag(override val text: String) : PostTextSegment

    data class Url(override val text: String, val url: String) : PostTextSegment
}

internal fun postTextSegments(value: String): List<PostTextSegment> = buildList {
    var cursor = 0
    POST_TOKEN_PATTERN.findAll(value).forEach { match ->
        if (match.range.first > cursor) appendPlain(value.substring(cursor, match.range.first))
        val token = match.value
        if (token.startsWith("#")) {
            add(PostTextSegment.Hashtag(token))
        } else {
            val link = token.trimUrlEnd()
            if (link.isSafeHttpUrl()) {
                add(PostTextSegment.Url(link, link))
                appendPlain(token.substring(link.length))
            } else {
                appendPlain(token)
            }
        }
        cursor = match.range.last + 1
    }
    appendPlain(value.substring(cursor))
}

private fun MutableList<PostTextSegment>.appendPlain(value: String) {
    if (value.isEmpty()) return
    val previous = lastOrNull()
    if (previous is PostTextSegment.Plain) {
        this[lastIndex] = PostTextSegment.Plain(previous.text + value)
    } else {
        add(PostTextSegment.Plain(value))
    }
}

private fun String.trimUrlEnd(): String {
    var end = length
    while (end > 0) {
        val final = this[end - 1]
        if (final in TRAILING_PUNCTUATION) {
            end -= 1
            continue
        }
        val candidate = substring(0, end)
        val unmatchedClosing = BRACKETS.any { (opening, closing) ->
            final == closing && candidate.count { it == closing } > candidate.count { it == opening }
        }
        if (unmatchedClosing) {
            end -= 1
            continue
        }
        break
    }
    return substring(0, end)
}

private fun String.isSafeHttpUrl(): Boolean = runCatching {
    val parsed = URI(this)
    (parsed.scheme.equals("http", ignoreCase = true) ||
        parsed.scheme.equals("https", ignoreCase = true)) && !parsed.host.isNullOrBlank()
}.getOrDefault(false)

private val POST_TOKEN_PATTERN = Regex(
    """https?://[^\s<>"'。、！？]+|#[\p{L}\p{N}_]+""",
    RegexOption.IGNORE_CASE,
)
private val TRAILING_PUNCTUATION = setOf('.', ',', '!', '?', ':', ';', '。', '、', '！', '？')
private val BRACKETS = listOf('(' to ')', '[' to ']', '{' to '}')

File: frontend/src/model/post-text-segments.ts
export type PostTextSegment =
  | { kind: "text" | "hashtag"; text: string }
  | { kind: "url"; text: string; url: string };

const TOKEN_PATTERN = /https?:\/\/[^\s<>"'。、！？]+|#[\p{L}\p{N}_]+/giu;
const TRAILING_PUNCTUATION = new Set([".", ",", "!", "?", ":", ";", "。", "、", "！", "？"]);
const BRACKETS = [
  ["(", ")"],
  ["[", "]"],
  ["{", "}"],
] as const;

export function postTextSegments(value: string): PostTextSegment[] {
  const segments: PostTextSegment[] = [];
  let cursor = 0;
  for (const match of value.matchAll(TOKEN_PATTERN)) {
    const start = match.index;
    const token = match[0];
    if (start > cursor) appendText(segments, value.slice(cursor, start));
    if (token.startsWith("#")) {
      segments.push({ kind: "hashtag", text: token });
    } else {
      const link = trimUrlEnd(token);
      if (isSafeHttpUrl(link)) {
        segments.push({ kind: "url", text: link, url: link });
        appendText(segments, token.slice(link.length));
      } else {
        appendText(segments, token);
      }
    }
    cursor = start + token.length;
  }
  appendText(segments, value.slice(cursor));
  return segments;
}

function appendText(segments: PostTextSegment[], value: string): void {
  if (value.length === 0) return;
  const last = segments.at(-1);
  if (last?.kind === "text") {
    last.text += value;
  } else {
    segments.push({ kind: "text", text: value });
  }
}

function trimUrlEnd(value: string): string {
  let end = value.length;
  while (end > 0) {
    const final = value[end - 1];
    if (final !== undefined && TRAILING_PUNCTUATION.has(final)) {
      end -= 1;
      continue;
    }
    const unmatchedClosing = BRACKETS.some(([opening, closing]) => {
      if (final !== closing) return false;
      const candidate = value.slice(0, end);
      return count(candidate, closing) > count(candidate, opening);
    });
    if (unmatchedClosing) {
      end -= 1;
      continue;
    }
    break;
  }
  return value.slice(0, end);
}

function count(value: string, token: string): number {
  return [...value].filter((character) => character === token).length;
}

function isSafeHttpUrl(value: string): boolean {
  try {
    const url = new URL(value);
    return (url.protocol === "http:" || url.protocol === "https:") && url.hostname.length > 0;
  } catch {
    return false;
  }
}
